crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -202,13 +217,26 @@ impl<'a> SenderDataFinder<'a> {
     }
 
     /// Step D (we have a device)
+    ///
+    /// Returns Err if the device does not own the session.
     fn have_device(&self, sender_device: Device) -> SenderData {
+        // Is the session owned by the device?
+        let device_is_owner = sender_device.is_owner_of_session(self.session);
+        if let Ok(false) | Err(_) = device_is_owner {
```

Comment:
Can we perhaps not use an early return here? I vastly prefer having each case listed in the same tree instead of having an early return I have to keep in mind. You can either match on both `device_is_owner` and `device_is_cross_signed()` or add an if clause to one of the match cases like so:

```rust
match device_is_owner {
    Ok(true) if sender_device.is_cross_signed_by_owner() => {
        // Is the device cross-signed?
        // Does the cross-signing key match that used to sign the device?
        // And is the signature in the device valid?
        self.device_is_cross_signed_by_sender(sender_device)
    }
    Ok(true) => {
        // No: F (we have device keys, but they are not signed by the sender)
        SenderData::DeviceInfo {
            device_keys: sender_device.as_device_keys().clone(),
            retry_details: SenderDataRetryDetails::retry_soon(),
            legacy_session: true, 
        }
    }
    Ok(false) | Err(_) => {
        // Step E (the device does not own the session)
        // Give up: something is wrong with the session.
        SenderData::UnknownDevice {
            retry_details: SenderDataRetryDetails::retry_soon(),
            legacy_session: true, // TODO: change to false when all SenderData work is done
            owner_check_failed: true,
        }
    }
}
```

Reply:
Good idea, fixed in db784d45d2